File: modules/libs/core/transcript/transcript.go
// Package transcript is what a model heard in a recording, written down as
// WebVTT.
//
//	WEBVTT
//
//	NOTE heard 9100
//
//	00:00:01.500 --> 00:00:04.200
//	what was said
//
//	00:00:04.200 --> 00:00:09.100
//	what was said next
//
// The format is the W3C one, so the file opens in a player, shows the words
// against the recording in a browser, and is read by anything a person already
// has. A run stopped part way says how far it got in a NOTE, which every reader
// of the format passes over.
//
// The timings are not the words. Reading takes them out, so an offset in what
// comes back is an offset in the speech, and a chunk cut from it holds what was
// said and not the bookkeeping around it.
//
// It is pure: no filesystem, no clock, no model.
package transcript

import (
	"bytes"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// Head is the line every WebVTT file begins with.
const Head = "WEBVTT"

// arrow separates the two timings of a cue.
const arrow = "-->"

// A Cue is one stretch of speech: what was said, when it was said, and where it
// stands in the words a transcript reads as. At is filled by reading, because
// only then is there a text for it to be an offset into.
type Cue struct {
	Text string
	From int
	To   int
	At   int
}

// Marshal is the artifact for a run of cues.
//
// A cue carrying no words is not written: silence is not something a person
// scrolls past, and a timing over nothing is a moment the recording never had.
func Marshal(cues []Cue) []byte {
	var out strings.Builder
	out.WriteString(Head)
	out.WriteString("\n")
	for _, cue := range cues {
		text := strings.TrimSpace(cue.Text)
		if text == "" {
			continue
		}
		fmt.Fprintf(&out, "\n%s %s %s\n%s\n", Stamp(cue.From), arrow, Stamp(cue.To), text)
	}
	return []byte(out.String())
}

// Parse is an artifact, as the words it holds and the cues they came from.
//
// The words of a cue stand one to a line, which is what a person reading the
// transcript sees and what a chunk is cut out of.
func Parse(raw []byte) (string, []Cue) {
	var out strings.Builder
	var cues []Cue
	for _, block := range strings.Split(strings.ReplaceAll(string(raw), "\r\n", "\n"), "\n\n") {
		cue, ok := parse(block)
		if !ok {
			continue
		}
		if out.Len() > 0 {
			out.WriteString("\n")
		}
		cue.At = out.Len()
		out.WriteString(cue.Text)
		cues = append(cues, cue)
	}
	return out.String(), cues
}

// parse is one block of the file as a cue, and whether it is one. The header,
// a note and anything a later version of the format adds are not.
func parse(block string) (Cue, bool) {
	lines := strings.Split(strings.Trim(block, "\n"), "\n")
	for at, line := range lines {
		before, after, found := strings.Cut(line, arrow)
		if !found {
			// A cue may be named on the line above its timing. Anything else
			// standing there is not a cue.
			if at > 0 || strings.HasPrefix(line, Head) || strings.HasPrefix(line, "NOTE") {
				return Cue{}, false
			}
			continue
		}
		from, ok := parseStamp(before)
		if !ok {
			return Cue{}, false
		}
		// Cue settings may follow the second timing, separated by a space. A
		// timing with nothing after the arrow is a line somebody was still
		// writing.
		second := strings.Fields(after)
		if len(second) == 0 {
			return Cue{}, false
		}
		to, ok := parseStamp(second[0])
		if !ok {
			return Cue{}, false
		}
		text := strings.TrimSpace(strings.Join(lines[at+1:], "\n"))
		if text == "" {
			return Cue{}, false
		}
		return Cue{Text: text, From: from, To: to}, true
	}
	return Cue{}, false
}

// Stamp is a millisecond as the format writes it: hours, minutes, seconds and
// thousandths.
func Stamp(ms int) string {
	if ms < 0 {
		ms = 0
	}
	return fmt.Sprintf("%02d:%02d:%02d.%03d", ms/3600000, ms/60000%60, ms/1000%60, ms%1000)
}

// Clock is a moment of a recording as a person reads one: the way a player
// writes where it stands. An hour that is not there is not written.
func Clock(ms int) string {
	whole := max(ms, 0) / 1000
	if hours := whole / 3600; hours > 0 {
		return fmt.Sprintf("%d:%02d:%02d", hours, whole/60%60, whole%60)
	}
	return fmt.Sprintf("%d:%02d", whole/60, whole%60)
}

// parseStamp is a timing the format writes. The hours are optional, which is
// what the format says and what other tools write.
func parseStamp(raw string) (int, bool) {
	parts := strings.Split(strings.TrimSpace(raw), ":")
	if len(parts) < 2 || len(parts) > 3 {
		return 0, false
	}
	ms := 0
	for _, part := range parts[:len(parts)-1] {
		n, err := strconv.Atoi(part)
		if err != nil {
			return 0, false
		}
		ms = ms*60 + n
	}
	ms *= 60000

	seconds, thousandths, found := strings.Cut(parts[len(parts)-1], ".")
	if !found {
		return 0, false
	}
	s, err := strconv.Atoi(seconds)
	if err != nil {
		return 0, false
	}
	t, err := strconv.Atoi(thousandths)
	if err != nil || len(thousandths) != 3 {
		return 0, false
	}
	return ms + s*1000 + t, true
}

// At is where a run of the words sits: the cues it falls in, in the order they
// were spoken. A run crossing a silence is in both of them.
func At(cues []Cue, start, length int) []Cue {
	if length <= 0 || len(cues) == 0 {
		return nil
	}
	end := start + length

	// The first cue that reaches into the run. A cue before it ends before the
	// run begins.
	at := sort.Search(len(cues), func(i int) bool {
		return cues[i].At+len(cues[i].Text) > start
	})

	var out []Cue
	for ; at < len(cues) && cues[at].At < end; at++ {
		out = append(out, cues[at])
	}
	return out
}

// Plays is the millisecond a run of the words is played from, and whether any
// cue holds it. A run no cue holds is nowhere to play.
func Plays(cues []Cue, start, length int) (int, bool) {
	found := At(cues, start, length)
	if len(found) == 0 {
		return 0, false
	}
	return found[0].From, true
}

// Heard is the note a run stopped part way leaves: how many milliseconds of the
// recording have been written down. It stands after the cues it claims, so a
// batch that did not land whole is one no note claims.
func Heard(ms int) []byte {
	return []byte(fmt.Sprintf("\nNOTE heard %d\n", ms))
}

// ByHand is the note a transcript a person wrote carries.
const ByHand = "NOTE by hand"

// Hand marks a transcript as the words a person put there. A transcript
// carrying it is left as they left it.
func Hand() []byte {
	return []byte("\n" + ByHand + "\n")
}

// Written says whether a person wrote these words. The mark is a note of its
// own, and the same words spoken in a cue are speech.
func Written(raw []byte) bool {
	for at := 0; at <= len(raw)-len(ByHand); {
		found := bytes.Index(raw[at:], []byte(ByHand))
		if found < 0 {
			return false
		}
		if found += at; begins(raw, found) {
			return true
		}
		at = found + 1
	}
	return false
}

// Reached is how far a run before this one got, and where the last note about
// it ends. A file carrying none is a recording nothing has listened to.
func Reached(raw []byte) (ms, end int) {
	note := []byte("NOTE heard ")
	// Each note is looked at once, and only what stands between it and the one
	// after it is read.
	for end := len(raw); ; {
		at := bytes.LastIndex(raw[:end], note)
		if at < 0 {
			return 0, 0
		}
		line := raw[at+len(note) : end]
		end = at
		stop := bytes.IndexByte(line, '\n')
		if stop < 0 || !begins(raw, at) {
			continue
		}
		ms, err := strconv.Atoi(strings.TrimSpace(string(line[:stop])))
		if err != nil {
			continue
		}
		return ms, at + len(note) + stop + 1
	}
}

// begins says whether a byte is where a block of the file starts: the top of
// it, or the line after a blank one. A note stands at the top of its own block.
func begins(raw []byte, at int) bool {
	if at == 0 {
		return true
	}
	if raw[at-1] != '\n' {
		return false
	}
	blank := at - 1
	if blank > 0 && raw[blank-1] == '\r' {
		blank--
	}
	return blank > 0 && raw[blank-1] == '\n'
}

File: modules/apps/desktop/editor/src/media/transcript.ts
/**
 * One recording as its tab reads it: where the player stands in it, its
 * transcript, which cue is being said now, and the words as a person edits
 * them.
 */
import type { Stretch } from '../shared/core'
import { computed, ref, shallowRef } from 'vue'
import { clock } from '@numen/ui'
import { troubleWords } from '@numen/wire'
import { answerGuard as latest } from '../shared/questions'
import { cued, same, spanning, spoken, type Cue } from './cues'
import { playable as canPlay, player, type MediaTypeProbe, type Player } from './player'
import { WORDS } from './words'

/** The player a tab drew for itself, which a moment chosen in the words seeks. */
export interface TabPlayer {
  seeks(ms: number): void
}

/** The text fetched or heard, and whether it may be written over. */
export interface Transcript {
  readonly cues: readonly Cue[]
  /**
   * The text where nothing timed it: the prose a page is written around. It
   * comes instead of the cues, never beside them.
   */
  readonly prose: string
  /** False while a run writing the transcript holds it. */
  readonly editable: boolean
}

/**
 * What a file is, for whatever plays it: how long it runs, in milliseconds,
 * and where its bytes come from.
 *
 * A recording nothing has listened to reaches nowhere, and the player it is
 * loaded into is what then says how long it runs.
 */
export interface RecordingSummary {
  readonly duration: number
  /** Where it is played from, as the application answers it. */
  readonly mediaUrl: string
  /**
   * What it is played as. The application says: what counts as a recording is
   * its to decide, and a url with no copy on this disk is a page to frame.
   */
  readonly mediaType: string
  /** The web address a url points at, and nothing on every other source. */
  readonly url: string
}

/** Everything a recording tab asks of the application. */
export interface Recordings {
  /** How long the recording runs, and how much of it has been written down. */
  listened(path: string): Promise<RecordingSummary>
  /** The words heard in the recording, in the order they were spoken. */
  cues(path: string): Promise<Transcript>
  /** The words as a person has edited them, kept against the recording. */
  writes(path: string, cues: readonly Cue[]): Promise<void>
  /**
   * The millisecond a stretch of the words written down is played from, and
   * nothing where no cue holds it.
   */
  plays(path: string, stretch: Stretch): Promise<number | null>
}

/**
 * The cue being said at a millisecond, and the last one said where a silence
 * stands there. Nothing until the first cue begins.
 */
const holding = (cues: readonly Cue[], ms: number): number => {
  for (let at = cues.length - 1; at >= 0; at--) {
    if (cues[at]!.from <= ms) return at
  }
  return -1
}

export type TranscriptState = ReturnType<typeof transcript>

/** How long the words have to have been still before they are written. */
export const QUIET = 800

/** What the window hands one recording tab, beside the application and the path. */
export interface TranscriptOptions {
  /** The player the sound comes out of, which every recording of a window shares. */
  through?: Player
  /** How long the typing settles for before the words are written. */
  quiet?: number
  /** Whether this window can play a kind of sound. */
  plays?: MediaTypeProbe
}

/** What a url with no copy on this disk is played as: a page, read in a frame. */
const PAGE = 'text/html'

export function transcript(recordings: Recordings, path: string, how: TranscriptOptions = {}) {
  const through = how.through ?? player
  const quiet = how.quiet ?? QUIET
  const plays = how.plays ?? canPlay()
  /** Where the recording's own bytes are played from, once it is asked. */
  const address = ref('')
  /** The words heard in the recording, in the order they were spoken. */
  const cues = shallowRef<readonly Cue[]>([])
  /** The words as the editor shows them, one cue to a line. */
  const prose = ref('')
  /** Whether the transcript may be written over now. */
  const editable = ref(true)
  /** Whether the view keeps the line being said in sight. */
  const following = ref(true)
  /** Whether a person has been typing too recently for the view to move. */
  const typing = ref(false)
  /** How long the recording runs, as the application last said. */
  const duration = ref(0)
  /** What the recording is played as, as the application answers it. */
  const type = ref('')
  /** The web address a url points at, and nothing on every other source. */
  const points = ref('')
  /** Whether the recording the player holds is this one. */
  const held = computed(() => address.value !== '' && through.address.value === address.value)

  /**
   * How long the recording runs. The application says, and the recording
   * itself says where it is loaded and knows better.
   */
  const runs = computed(() => Math.max(duration.value, held.value ? through.duration.value : 0))

  /**
   * Where the player stands in this recording, in milliseconds.
   *
   * The window plays one recording at a time, so one the player is not holding
   * stands at its beginning until somebody plays it.
   */
  const now = computed(() => {
    if (frame.value) return Math.max(0, framed.value)
    return held.value ? through.at.value : 0
  })

  /**
   * The player drawn in the tab, which a url has and a recording does not: what
   * a url points at plays where it is drawn, not through the one the window
   * plays sound with.
   */
  const frame = shallowRef<TabPlayer | null>(null)
  /** Where that player stands, and nothing before it has said. */
  const framed = ref(-1)

  /** The tab hands over the player it drew, and says where it stands. */
  const playsIn = (player: TabPlayer | null) => {
    frame.value = player
    framed.value = -1
  }
  const reached = (ms: number) => {
    framed.value = ms
  }

  /** Whether this recording is the one playing. */
  const playing = computed(() => held.value && through.playing.value)
  /** Whether something is writing down what this recording says. */
  const working = ref(false)
  /** What this recording could not do, in words the tab puts up for it. */
  const trouble = ref('')
  /** What the player could not do, while this is the recording it holds. */
  const broken = computed(() => (held.value ? through.failed.value : ''))

  /**
   * The lines on screen against the milliseconds they cover. The lines are
   * what a person edits, and these follow them until the file is written.
   *
   * A recording nothing was heard in and nothing was typed into has no lines
   * at all, and the tab says so where they would stand.
   */
  const spans = computed(() =>
    cues.value.length === 0 && prose.value === '' ? [] : spanning(cues.value, prose.value),
  )

  /**
   * Whether the text carries the times each stretch of it was said at. A page
   * is prose and carries none: there is nothing to seek and no line being said.
   */
  const timed = computed(() => cues.value.length > 0)

  /** Whether anything at all has been fetched or heard here. */
  const written = computed(() => timed.value || prose.value !== '')

  /** Which line is being said now, and nothing where none has begun. */
  const current = computed(() =>
    frame.value && framed.value < 0 ? -1 : holding(spans.value, now.value),
  )

  /** Whether the tab this recording stands in is still open. */
  let open = true

  /**
   * A moment gone to before the recording knew its own address, played from
   * once it does. A hit in the words opens a tab and asks for a moment in the
   * same breath.
   */
  let wanted = -1

  /**
   * A run being written down is asked about again while it goes, and two
   * answers may arrive in either order. The older of them carries fewer words.
   */
  const asks = latest()

  /** Whether what is on screen has still to reach the file. */
  let owed = false
  /** A write of the words that has not answered yet. */
  let writing = false
  /** The wait the typing is being let settle over. */
  let settling: ReturnType<typeof setTimeout> | undefined
  /** The wait after which the view may go after the words again. */
  let stilling: ReturnType<typeof setTimeout> | undefined

  /**
   * What the recording is and what has been written down of it. A build that
   * cannot read a transcript says so where the words would stand, and the
   * recording still plays.
   */
  const reads = async () => {
    const mine = asks.ask()
    try {
      const said = await recordings.listened(path)
      if (!mine.lands()) return
      duration.value = said.duration
      address.value = said.mediaUrl
      type.value = said.mediaType
      points.value = said.url
      // A moment asked for before the recording knew where its bytes are.
      if (wanted >= 0 && address.value) {
        const at = wanted
        wanted = -1
        through.seek(address.value, at)
      }
      // The player holding nothing takes this recording, so the controls read
      // how long it runs before anybody presses play. One already in the
      // player is left where it is, and a url plays where it is drawn.
      if (!points.value && address.value && through.address.value === '') {
        through.load(address.value)
      }

      const spoke = await recordings.cues(path)
      if (!mine.lands()) return
      cues.value = spoke.cues
      editable.value = spoke.editable
      // Words the person has typed and not yet had written stay on screen.
      if (!owed) prose.value = spoke.cues.length ? spoken(spoke.cues) : spoke.prose
      trouble.value = ''
    } catch (error) {
      if (!mine.lands()) return
      trouble.value = troubleWords(error)
    }
  }

  /** What the recording is, asked for as its tab opens. */
  const opened = reads()

  /** The moment the person went to. Before the beginning is the beginning. */
  const go = (ms: number) => {
    if (!open) return
    const at = Math.max(0, Math.round(ms))
    if (frame.value) return frame.value.seeks(at)
    if (!address.value) return void (wanted = at)
    through.seek(address.value, at)
  }

  /** Play this recording, taking the sound from whatever else held it. */
  const play = () => {
    if (!open || !address.value) return
    through.play(address.value)
  }

  /** Stop it, while it is this recording that is playing. */
  const pause = () => {
    if (playing.value) through.pause()
  }

  /** The line a person asked for, counted from the first line on screen. */
  const goes = (line: number) => {
    const span = spans.value[line]
    if (span) go(span.from)
  }

  /** Whether the view keeps the line being said in sight. */
  const follows = (on: boolean) => {
    following.value = on
  }

  /** The person typed. The words are written once they have been still. */
  const typed = (body: string) => {
    if (!open || body === prose.value) return
    prose.value = body
    owed = true
    // An edit that adds or takes away a line moves which line is being said,
    // and the view does not go after a line a person moved under their own
    // hands.
    typing.value = true
    clearTimeout(stilling)
    stilling = setTimeout(() => void (typing.value = false), quiet)
    clearTimeout(settling)
    settling = setTimeout(() => void keep(), quiet)
  }

  /**
   * The words as they now read, kept against the recording. A write that is
   * refused leaves them owed, so the next stillness offers them again.
   */
  const keep = async () => {
    clearTimeout(settling)
    settling = undefined
    if (!open || !owed || writing || !editable.value) return
    const body = prose.value
    const next = cued(cues.value, body)
    owed = false
    // A transcript written down is a transcript a person owns, and a
    // proofreader leaves it alone. Only words that changed are written.
    if (same(next, cues.value)) return
    writing = true
    try {
      await recordings.writes(path, next)
      if (!open) return
      cues.value = next
      trouble.value = ''
    } catch (error) {
      if (!open) return
      owed = true
      trouble.value = troubleWords(error)
    } finally {
      writing = false
      // Typing that landed while the write was in the air is still owed.
      if (open && prose.value !== body) {
        owed = true
        settling = setTimeout(() => void keep(), quiet)
      }
    }
  }

  /**
   * The words are asked for again, and what stands on screen is whatever comes
   * back. An edit still waiting to be written goes: it was of words that are no
   * longer the ones the recording has.
   */
  const again = () => {
    clearTimeout(settling)
    settling = undefined
    owed = false
    void reads()
  }

  /**
   * Work on this recording, as the application last reported it. The words are
   * asked for again while a run is going and once more when it stops.
   */
  const ticks = (running: boolean) => {
    if (!open || (!running && !working.value)) return
    working.value = running
    void reads()
  }

  /**
   * Stretches of the words written down reached: the player is sent to the
   * moment the first of them was spoken at. A stretch no cue holds leaves the
   * player where it stands.
   */
  const reach = async (...stretches: readonly Stretch[]) => {
    await opened
    if (!open || stretches.length === 0) return
    try {
      const ms = await recordings.plays(path, stretches[0]!)
      if (!open || ms === null) return
      go(ms)
    } catch (error) {
      if (!open) return
      trouble.value = troubleWords(error)
    }
  }

  /**
   * The tab has closed: what the person typed reaches the file, nothing is
   * asked for again, and the recording stops where the player stands in it.
   *
   * Two tabs may stand on one recording, and closing either of them stops it.
   */
  const close = () => {
    void keep()
    pause()
    open = false
    asks.close()
    clearTimeout(stilling)
    cues.value = []
    prose.value = ''
  }

  /**
   * The moment each line was said at, on a clock, as the editor's gutter draws
   * them. These follow the words alone, so a transcript of any length is
   * written out once and left alone while the recording plays.
   */
  const times = computed(() => spans.value.map((cue) => clock(cue.from)))

  /** Whether this window can play a recording of this kind at all. */
  const playable = computed(() => address.value !== '' && plays(type.value))

  /**
   * Whether what is at the address is a page, which is read in a frame rather
   * than played. A url with no copy on this disk is one.
   */
  const framing = computed(() => address.value !== '' && type.value === PAGE)

  /** How far into the recording the words written down reach, in milliseconds. */
  const transcribedDuration = computed(() => cues.value.at(-1)?.to ?? 0)

  /** What the tab says where the words would stand, and nothing where they do. */
  const note = computed(() => {
    if (times.value.length) return ''
    if (working.value) return WORDS.transcribing
    return points.value ? WORDS.unfetched : WORDS.silence
  })

  return {
    path,
    address,
    points,
    framing,
    playable,
    times,
    spans,
    note,
    cues,
    prose,
    editable,
    following,
    typing,
    duration,
    runs,
    transcribedDuration,
    now,
    current,
    timed,
    written,
    playsIn,
    reached,
    working,
    trouble,
    broken,
    go,
    goes,
    follows,
    typed,
    keep,
    again,
    playing,
    play,
    pause,
    ticks,
    reach,
    close,
  }
}
